### PR TITLE
Upgrade to ASM 5

### PR DESCRIPTION
**Pull request by ebourg**
Hi,

I'd like to suggest this modification to upgrade to the latest version of ASM.

Thank you

### CODE REVIEW

**Comment by Batmat** [comment]
Unfortunately the PR failed, but I guess this is a good idea and a condition (?) to be able to compile a jenkins plugin with a JDK8.

**Comment by ebourg** [comment]
The build failure reported by buildhive was caused by a missing artifact in Maven central (annotation-indexer 1.4). I don't think the changes I suggested are responsible for this error.

**Comment by Batmat** [comment]
I guess that would be better to make it pass indeed. Either have @kohsuke retrigger it or by you adding a commit (yeah that sounds ugly)

**Comment by ebourg** [comment]
annotation-indexer 1.4 hasn't been deployed to Maven central yet, would @kohsuke be able to do it? If it's available from an alternative repository I may also add it to the pom.

**Comment by Batmat** [comment]
So I think the version should first be deployed to Central. If it's not
deployed in Central, it's not :).
Le 12 févr. 2015 14:02, "Emmanuel Bourg" <notifications@github.com> a écrit
:

**Comment by kohsuke** [comment]
annotation-indexer 1.4 is posted to central now.

**Comment by ebourg** [comment]
Thanks!

**Comment by Batmat** [comment]
+1, thanks

File: access-modifier-checker/src/main/java/org/kohsuke/accmod/impl/Checker.java [hunk 1]
```diff
@@ -163,7 +163,7 @@ public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
 
                     @Override
                     public MethodVisitor visitMethod(int access, final String methodName, final String methodDesc, String signature, String[] exceptions) {
-                        return new EmptyVisitor() {
+                        return new MethodVisitor(Opcodes.ASM5) {
```
Comment (jglick) [inline access-modifier-checker/src/main/java/org/kohsuke/accmod/impl/Checker.java]: AFAICT this completely disabled `visitMethodInsn` and thus broke the most common use case of this library. Unfortunately it has no test suite…

Comment (ebourg) [inline access-modifier-checker/src/main/java/org/kohsuke/accmod/impl/Checker.java]: I'm not sure to understand, in ASM 3.2 `EmptyVisitor` has [an empty implementation](http://websvn.ow2.org/blame.php?repname=asm&path=%2Ftags%2FASM_3_2%2Fsrc%2Forg%2Fobjectweb%2Fasm%2Fcommons%2FEmptyVisitor.java&rev=1389#l156) for `visitMethodInsn`. If the method was called it did nothing. In ASM 5.x `MethodVisitor.visitMethodInsn()` has a [delegation mechanism](http://websvn.ow2.org/blame.php?repname=asm&path=%2Ftags%2FASM_5_2%2Fsrc%2Forg%2Fobjectweb%2Fasm%2FMethodVisitor.java&rev=1841#l456) but it isn't used in this context, so the method also does nothing. What did I miss?

Comment (ebourg) [inline access-modifier-checker/src/main/java/org/kohsuke/accmod/impl/Checker.java]: Actually I wasn't looking at the right line, the `MethodVisitor` at line 246 implements `visitMethodInsn`, not the one at line 166. That said, it's still not clear why it breaks. In ASM 5.2 `visitMethodInsn(int, String, String String)` calls `visitMethodInsn(int, String, String String, boolean)` when the api level used is `Opcodes.ASM5`. So even if the deprecated `visitMethodInsn` method is used it should still work.

**Pull request merged**